sidechain/block-composer/src/block_composer.rs

```diff
@@ -37,133 +36,106 @@ use std::{format, marker::PhantomData, sync::Arc, vec::Vec};
 
 /// Compose a sidechain block and corresponding confirmation extrinsic for the parentchain
 ///
-pub trait ComposeBlockAndConfirmation {
+pub trait ComposeBlockAndConfirmation<Externalities, PB: BlockT> {
 	type SidechainBlockT: SignedBlockT;
-	type ParentchainBlockT: BlockT;
 
 	fn compose_block_and_confirmation(
 		&self,
-		latest_onchain_header: &<Self::ParentchainBlockT as BlockT>::Header,
+		latest_parentchain_header: &<PB as BlockT>::Header,
 		top_call_hashes: Vec<H256>,
 		shard: ShardIdentifier,
 		state_hash_apriori: H256,
+		aposteriori_state: Externalities,
 	) -> Result<(OpaqueCall, Self::SidechainBlockT)>;
 }
 
 /// Block composer implementation for the sidechain
-pub struct BlockComposer<PB, SB, Signer, StateKey, RpcAuthor, StfExecutor> {
+pub struct BlockComposer<PB, SB, Signer, StateKey, RpcAuthor> {
 	signer: Signer,
 	state_key: StateKey,
 	rpc_author: Arc<RpcAuthor>,
-	stf_executor: Arc<StfExecutor>,
 	_phantom: PhantomData<(PB, SB)>,
 }
 
-impl<PB, SB, Signer, StateKey, RpcAuthor, StfExecutor>
-	BlockComposer<PB, SB, Signer, StateKey, RpcAuthor, StfExecutor>
+impl<PB, SB, Signer, StateKey, RpcAuthor> BlockComposer<PB, SB, Signer, StateKey, RpcAuthor>
 where
 	PB: BlockT<Hash = H256>,
 	SB: SignedBlockT<Public = Signer::Public, Signature = MultiSignature>,
 	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
 	SB::Signature: From<Signer::Signature>,
 	RpcAuthor:
 		AuthorApi<H256, PB::Hash> + OnBlockCreated<Hash = PB::Hash> + SendState<Hash = PB::Hash>,
-	StfExecutor: StfExecuteGenericUpdate,
-	<StfExecutor as StfExecuteGenericUpdate>::Externalities:
-		SgxExternalitiesTrait + SidechainState + SidechainSystemExt + StateHash,
 	Signer: Pair<Public = sp_core::ed25519::Public>,
 	Signer::Public: Encode,
 	StateKey: StateCrypto,
 {
-	pub fn new(
-		signer: Signer,
-		state_key: StateKey,
-		rpc_author: Arc<RpcAuthor>,
-		stf_executor: Arc<StfExecutor>,
-	) -> Self {
-		BlockComposer { signer, state_key, rpc_author, stf_executor, _phantom: Default::default() }
+	pub fn new(signer: Signer, state_key: StateKey, rpc_author: Arc<RpcAuthor>) -> Self {
+		BlockComposer { signer, state_key, rpc_author, _phantom: Default::default() }
 	}
 }
 
-impl<PB, SB, Signer, StateKey, RpcAuthor, StfExecutor> ComposeBlockAndConfirmation
-	for BlockComposer<PB, SB, Signer, StateKey, RpcAuthor, StfExecutor>
+impl<PB, SB, Signer, StateKey, RpcAuthor, Externalities>
+	ComposeBlockAndConfirmation<Externalities, PB>
+	for BlockComposer<PB, SB, Signer, StateKey, RpcAuthor>
 where
 	PB: BlockT<Hash = H256>,
 	SB: SignedBlockT<Public = Signer::Public, Signature = MultiSignature>,
 	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
 	SB::Signature: From<Signer::Signature>,
 	RpcAuthor:
 		AuthorApi<H256, PB::Hash> + OnBlockCreated<Hash = PB::Hash> + SendState<Hash = PB::Hash>,
-	StfExecutor: StfExecuteGenericUpdate,
```

Comment:
No stf executor interaction anymore. Simply take the state as input value and compose the block with this state.

Reply:
This seems like a nice simplification and more in-line with what one would expect from a 'block composer' 👍